src/tracer-config.ts

```diff
@@ -0,0 +1,159 @@
+
+  // Merge the environments
+  const env: { [key: string]: string; } = {};
+  let copyExecutables = false;
+  let envSize = 0;
+  for (const v of Object.values(tracerConfigs)) {
+    for (let e of Object.entries(v.env)) {
+      const name = e[0];
+      const value = e[1];
+      // skip SEMMLE_COPY_EXECUTABLES_ROOT as it is handled separately
+      if (name === 'SEMMLE_COPY_EXECUTABLES_ROOT') {
+        copyExecutables = true;
+      } else if (name in env) {
+        if (env[name] !== value) {
+          throw Error('Incompatible values in environment parameter ' +
+            name + ': ' + env[name] + ' and ' + value);
+        }
+      } else {
+        env[name] = value;
+        envSize += 1;
+      }
+    }
+  }
+
+  // Concatenate spec files into a new spec file
+  let languages = Object.keys(tracerConfigs);
+  const cppIndex = languages.indexOf('cpp');
+  // Make sure cpp is the last language, if it's present since it must be concatenated last
+  if (cppIndex !== -1) {
+    let lastLang = languages[languages.length - 1];
+    languages[languages.length - 1] = languages[cppIndex];
+    languages[cppIndex] = lastLang;
+  }
+
+  let totalLines: string[] = [];
+  let totalCount = 0;
+  for (let lang of languages) {
+    const lines = fs.readFileSync(tracerConfigs[lang].spec, 'utf8').split(/\r?\n/);
+    const count = parseInt(lines[1], 10);
+    totalCount += count;
+    totalLines.push(...lines.slice(2));
+  }
+
+  const newLogFilePath = path.resolve(config.tempDir, 'compound-build-tracer.log');
+  const spec = path.resolve(config.tempDir, 'compound-spec');
+  const compoundTempFolder = path.resolve(config.tempDir, 'compound-temp');
+  const newSpecContent = [newLogFilePath, totalCount.toString(10), ...totalLines];
+
+  if (copyExecutables) {
+    env['SEMMLE_COPY_EXECUTABLES_ROOT'] = compoundTempFolder;
+    envSize += 1;
+  }
+
+  fs.writeFileSync(spec, newSpecContent.join('\n'));
+
+  // Prepare the content of the compound environment file
+  let buffer = Buffer.alloc(4);
+  buffer.writeInt32LE(envSize, 0);
+  for (let e of Object.entries(env)) {
+    const key = e[0];
+    const value = e[1];
+    const lineBuffer = new Buffer(key + '=' + value + '\0', 'utf8');
+    const sizeBuffer = Buffer.alloc(4);
+    sizeBuffer.writeInt32LE(lineBuffer.length, 0);
+    buffer = Buffer.concat([buffer, sizeBuffer, lineBuffer]);
+  }
+  // Write the compound environment
+  const envPath = spec + '.environment';
+  fs.writeFileSync(envPath, buffer);
+
+  return { env, spec };
+}
+
+export async function getTracerConfig(
+  config: configUtils.Config,
+  codeql: CodeQL): Promise<TracerConfig | undefined> {
+
+  // Abort if there are no traced languages as there's nothing to do
+  const tracedLanguages = config.languages.filter(isTracedLanguage);
+  if (tracedLanguages.length === 0) {
+    return undefined;
+  }
+
+  // Get all the tracer configs and combine them together
+  let tracedLanguageConfigs: { [lang: string]: TracerConfig } = {};
```

Comment:
I think this can be a constants as you're not reassigning it, just updating its inner values.

```suggestion
  const tracedLanguageConfigs: { [lang: string]: TracerConfig } = {};
```